juspay/hyperswitch
# refactor: add basic counter metrics for IMC

## Type of Change



## Description

This PR introduces basic counter metrics for IMC -

* hits (successful read)
* misses (failed read)
* manual invalidations (done by applications) - makes use of eviction listener on moka that can help identify strategy based evictions

This PR also moves the `add_attribute` method of metrics to `router_env` crate. This method now takes a list of attributes as an argument as opposed to just one (key,value) pair.

### Additional Changes





## Motivation and Context



## How did you test it?

1. Insert new config key value pair. Data is not cached in IMC during insertion

```
curl --location 'http://localhost:8080/configs/' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "key": "test_key",
    "value": "test_val_1"
}'
```

2. Retrieve the inserted config so that the data gets cached to IMC. Since this is the first time the data is getting retrieved, there'll be a cache mis

```
curl --location 'http://localhost:8080/configs/test_key' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data ''
```

![image](https://github.com/juspay/hyperswitch/assets/70657455/99f05a27-3ddd-4e5d-9458-f5673423ddc8)

3.  Again retrieve the same config. Since in the previous retrieve, the data was cached, this retrieve will be a cache hit

```
curl --location 'http://localhost:8080/configs/test_key' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data ''
```

![image](https://github.com/juspay/hyperswitch/assets/70657455/b3ecd6da-3268-4830-be0b-7ca677f6c4d4)

4. Update the inserted config. This time the cache data has to be invalidated. This should increment the `IN_MEMORY_CACHE_EVICTION_COUNT` metric

```
curl --location 'http://localhost:8080/configs/test_key' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "key": "test_key",
    "value": "test_val_2"
}'
```

![image](https://github.com/juspay/hyperswitch/assets/70657455/d9492d34-9767-4e41-83a0-c0462af047b5)


## Checklist
